txcache/eviction.go

```diff
@@ -67,62 +82,94 @@ func (cache *TxCache) areThereTooManyTxs() bool {
+	for i, bunch := range bunches {
+		if len(bunch) == 0 {
+			// Some senders may have no transaction anymore (hazardous concurrent removals).
+			continue
+		}
+
+		// Items will be reused (see below). Each sender gets one (and only one) item in the heap.
+		heap.Push(&transactionsHeap, &TransactionsHeapItem{
+			senderIndex:      i,
+			transactionIndex: 0,
+			transaction:      bunch[0],
+		})
 	}
 
-	snapshot := cache.evictionSnapshotOfSenders
-	snapshotLength := uint32(len(snapshot))
-	batchSize := cache.config.NumSendersToPreemptivelyEvict
-	batchStart := uint32(0)
+	for pass := 0; cache.isCapacityExceeded(); pass++ {
+		transactionsToEvict := make(BunchOfTransactions, 0, cache.config.NumItemsToPreemptivelyEvict)
+		transactionsToEvictHashes := make([][]byte, 0, cache.config.NumItemsToPreemptivelyEvict)
+
+		// Select transactions (sorted).
+		for transactionsHeap.Len() > 0 {
+			// Always pick the "worst" transaction.
+			item := heap.Pop(&transactionsHeap).(*TransactionsHeapItem)
 
-	for step = 0; shouldContinue(); step++ {
-		batchEnd := batchStart + batchSize
-		batchEndBounded := core.MinUint32(batchEnd, snapshotLength)
-		batch := snapshot[batchStart:batchEndBounded]
+			if len(transactionsToEvict) >= int(cache.config.NumItemsToPreemptivelyEvict) {
+				// We have enough transactions to evict in this pass.
+				break
+			}
 
-		numTxsEvictedInStep, numSendersEvictedInStep := cache.evictSendersAndTheirTxs(batch)
+			transactionsToEvict = append(transactionsToEvict, item.transaction)
+			transactionsToEvictHashes = append(transactionsToEvictHashes, item.transaction.TxHash)
 
-		numTxs += numTxsEvictedInStep
-		numSenders += numSendersEvictedInStep
-		batchStart += batchSize
+			// If there are more transactions in the same bunch (same sender as the popped item),
+			// add the next one to the heap (to compete with the others in being "the worst").
+			item.transactionIndex++
 
-		reachedEnd := batchStart >= snapshotLength
-		noTxsEvicted := numTxsEvictedInStep == 0
-		incompleteBatch := numSendersEvictedInStep < batchSize
+			if item.transactionIndex < len(bunches[item.senderIndex]) {
+				// Item is reused (same originating sender), pushed back on the heap.
+				item.transaction = bunches[item.senderIndex][item.transactionIndex]
+				heap.Push(&transactionsHeap, item)
+			}
+		}
 
-		shouldBreak := noTxsEvicted || incompleteBatch || reachedEnd
-		if shouldBreak {
+		if len(transactionsToEvict) == 0 {
+			// No more transactions to evict.
 			break
 		}
-	}
 
-	return
-}
+		// For each sender, find the "lowest" (in nonce) transaction to evict.
+		lowestToEvictBySender := make(map[string]uint64)
+
+		for _, tx := range transactionsToEvict {
+			sender := string(tx.Tx.GetSndAddr())
+			lowestToEvictBySender[sender] = tx.Tx.GetNonce()
+		}
+
+		// Remove those transactions from "txListBySender".
+		for sender, nonce := range lowestToEvictBySender {
+			cache.txListBySender.evictTransactionsWithHigherOrEqualNonces([]byte(sender), nonce)
```

Comment:
it seems like here there is some double work, if the same sender is selected 2 times in a capacity exceeded event.

Reply:
Before the loop that calls `evictTransactionsWithHigherOrEqualNonces` for each sender, there is another loop over the `transactionsToEvict`, where we retain the lowest nonces (to evict), by sender. Thus, within an _eviction pass_, we call `evictTransactionsWithHigherOrEqualNonces` only once for a given sender.

However, you are right, an eviction can happen in one or more passes. Currently, we define an _eviction pass_ as `pick worst transactions & actually remove them`, not as `accumulate worst transactions to be removed at the end of all passes`. This has the advantage to not require a rework of "isCapacityExceeded" (which currently operates on the actual counts of transactions, not of "virtual" / "to-be" counts). Additionally, it's also good to remove the transactions quickly after deciding they should be evicted (thus, not accumulating them across all eviction passes, then wait until the last pass ends).

An additional note: since the number `NumItemsToPreemptivelyEvict` is set to a large value in practice (50k transactions, set in _mx-chain-go_), most evictions will happen in one pass.

All right as it is now?